### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by physikerwelt**
mocha tests for MathJax-node

### CODE REVIEW

**Comment by physikerwelt** [comment]
See https://coveralls.io/r/physikerwelt/MathJax-node for the automated runs https://travis-ci.org/physikerwelt/MathJax-node

**Comment by physikerwelt** [comment]
This would close #83

**Comment by physikerwelt** [comment]
@dpvc What do you think about automated tests for MathJax-node?

**Comment by physikerwelt** [comment]
ping

**Comment by pkra** [comment]
Thanks for this! We are all busy on other parts of MathJax right now but I'll bring it up on our next F2F.

**Comment by physikerwelt** [comment]
I discussed this issue with @zorkow. I need to update the asciimath test.

**Comment by physikerwelt** [comment]
@pkra see https://travis-ci.org/physikerwelt/MathJax-node/builds/71753279

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ batik/*
 !batik/README.md
 mathjax/*
 !mathjax/README.md
+MathJax-node.iml
```
Comment (pkra) [inline .gitignore]: What's the purpose of this?

File: .jshintignore [hunk 2]
```diff
@@ -0,0 +1,4 @@
+coverage
```
Comment (pkra) [inline .jshintignore]: This seem unnecessary since we're not enforcing coding styles.

Comment (pkra) [inline .jshintignore]: (I mean the entire file)

File: test/README.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+                    var fs = require('fs');
+                    fs.writeFile("/vagrant/srv/test", JSON.stringify(res), function(err) {
+                        if(err) {
+                            return console.log(err);
+                        }
+                        console.log("The file was saved!");
+                    });
+```
+
+In .jsignore lib is ignored. That should be changed in the future
```
Comment (pkra) [inline test/README.md]: Again, we don't enforce coding styles.

File: .gitignore (hunk 1 above)
Comment (physikerwelt) [inline .gitignore]: I can remove that... it's just the intellij project

File: .jshintignore (hunk 2 above)
Comment (physikerwelt) [inline .jshintignore]: ok we can remove that

File: test/README.md (hunk 3 above)
Comment (physikerwelt) [inline test/README.md]: ok.

File: test-files/sample-mml.html [hunk 4]
```diff
@@ -1,116 +1,116 @@
...
-
-    <style>#content { margin: 0 0 36px 0; }</style>
-
-	</div><!-- #main -->
-
-</div><!-- #wrapper -->
-
-</body>
-</html>
+<!DOCTYPE html>
```
Comment (pkra) [inline test-files/sample-mml.html]: What's the purpose of this file? I don't see it used anywhere. (It seems to be just a copy of a page from the old MathJax wordpress site?)

File: test/README.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+== How to generate tests ==
+
+... to be written ...
+```
+                    var fs = require('fs');
```
Comment (pkra) [inline test/README.md]: Maybe drop lines 5++?

File: test/features/math/simple.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            response: {
+                "mml": "<math xmlns=\"http://www.w3.org/1998/Math/MathML\" display=\"block\" alttext=\"upper E equals m c squared\">\n  <semantics>\n    <mrow>\n      <mi>E</mi>\n      <mo>=</mo>\n      <mi>m</mi>\n      <msup>\n        <mi>c</mi>\n        <mn>2</mn>\n      </msup>\n    </mrow>\n    <annotation encoding=\"application/x-tex\">E=mc^2</annotation>\n  </semantics>\n</math>",
+                "speakText": "upper E equals m c squared"
+            }
+        },
+        {
+            query: {
+                math: "\\overline{A}^T",
+                mml: true,
+                speakText: true
```
Comment (pkra) [inline test/features/math/simple.js]: See below.

File: test/features/math/simple.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+var testData = [
+        {
+            query: {
+                math: 'E=mc^2',
+                mml: true,
+                speakText: true
+            },
+            response: {
+                "mml": "<math xmlns=\"http://www.w3.org/1998/Math/MathML\" display=\"block\" alttext=\"upper E equals m c squared\">\n  <semantics>\n    <mrow>\n      <mi>E</mi>\n      <mo>=</mo>\n      <mi>m</mi>\n      <msup>\n        <mi>c</mi>\n        <mn>2</mn>\n      </msup>\n    </mrow>\n    <annotation encoding=\"application/x-tex\">E=mc^2</annotation>\n  </semantics>\n</math>",
+                "speakText": "upper E equals m c squared"
```
Comment (pkra) [inline test/features/math/simple.js]: I'm not sure about this. It seems to me, tests should be done as far upstream as possible, i.e., in this case at  speechruleengine.

Comment (pkra) [inline test/features/math/simple.js]: This affects line 11 and the same in the other tests

File: test/features/math/simple.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+var testData = [
+        {
+            query: {
+                math: 'E=mc^2',
+                mml: true,
+                speakText: true
+            },
+            response: {
+                "mml": "<math xmlns=\"http://www.w3.org/1998/Math/MathML\" display=\"block\" alttext=\"upper E equals m c squared\">\n  <semantics>\n    <mrow>\n      <mi>E</mi>\n      <mo>=</mo>\n      <mi>m</mi>\n      <msup>\n        <mi>c</mi>\n        <mn>2</mn>\n      </msup>\n    </mrow>\n    <annotation encoding=\"application/x-tex\">E=mc^2</annotation>\n  </semantics>\n</math>",
```
Comment (pkra) [inline test/features/math/simple.js]: This seems to be comparing strings. That's ok for. 

A better way would be XML/HTML/DOM diffing; otherwise we might get unnecessary failing tests, e.g., when attribute order changes.

I suppose TeX-to-MathML tests really belong upstream with MathJax. Then again, they fit much better in mathjax-node. Hm...

File: test/features/math/simple.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+                "speakText": "x squared or a Subscript m n Baseline or a Subscript m n Baseline or StartFraction x plus 1 Over y EndFraction or StartRoot x EndRoot"
+           }
+        }
+
+    ]
+    ;
+
+
+describe('Simple Mathoid API tests', function () {
```
Comment (pkra) [inline test/features/math/simple.js]: This is Mathoid specific?

File: test-files/sample-mml.html (hunk 4 above)
Comment (physikerwelt) [inline test-files/sample-mml.html]: Something must be changed. Maybe line-breaks, or encoding... I can not tell from the git ui... but I did not change anything intentional.

File: test/README.md (hunk 5 above)
Comment (physikerwelt) [inline test/README.md]: Yes it might be hard to understand without context...

File: test/features/math/simple.js (hunk 8 above)
Comment (physikerwelt) [inline test/features/math/simple.js]: Are there automatic tests for mathjax... I think it would be best to try first some inital tests with mathjaxnode and push up to mathjax thereafter. @pkra I predict that you'll enjoy automatic tests, once you get used to them.

File: test/features/math/simple.js (hunk 7 above)
Comment (physikerwelt) [inline test/features/math/simple.js]: I can git rid of everything you don't like so we can start with one really simple test.

File: test/features/math/simple.js (hunk 9 above)
Comment (physikerwelt) [inline test/features/math/simple.js]: no.

**Comment by physikerwelt** [comment]
@pkra Can you summarize what you think should be removed. I have no reservations to start with a really simple first test.

File: test/features/math/simple.js (hunk 7 above)
Comment (pkra) [inline test/features/math/simple.js]: I'm sorry, I didn't meant to be overly critical. Let's discuss it F2F later.

File: test/features/math/simple.js (hunk 8 above)
Comment (pkra) [inline test/features/math/simple.js]: > Are there automatic tests for mathjax

Of course there are -- well over 1000. https://github.com/mathjax/MathJax-test/tree/master/testsuite

> I predict that you'll enjoy automatic tests, once you get used to them.

We have been used to them for a few years :-) We would love to refactor the test suite when we have the time (yeah right...), which includes separating the pure JS tests from the Selenium tests.

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+language: node_js
+node_js:
+    - "0.10"
+    - "0.11"
+    - "0.12"
+    - "iojs"
+
+script: npm run-script coverage && (npm run-script coveralls || exit 0)
```
Comment (pkra) [inline .travis.yml]: As per F2F, let's skip coveralls for now.

File: test-files/sample-mml.html (hunk 4 above)
Comment (pkra) [inline test-files/sample-mml.html]: My bad -- this file was not new. Let's drop it.

Comment (pkra) [inline test-files/sample-mml.html]: by which I mean: discuss.... @dpvc is this referenced anywhere? mjpage perhaps?

File: package.json [hunk 11]
```diff
@@ -23,5 +23,18 @@
...
+      "assert": "^1.3.0",
+      "istanbul": "^0.3.13",
+      "mocha": "^2.2.4",
+      "mocha-jshint": "0.0.9",
+      "mocha-lcov-reporter": "0.0.1",
+      "coveralls": "2.11.2"
+    },
+    "scripts": {
+      "test": "mocha",
+      "coverage": "istanbul cover _mocha -- -R spec",
```
Comment (physikerwelt) [inline package.json]: Delete

File: package.json [hunk 12]
```diff
@@ -23,5 +23,18 @@
...
+      "coveralls": "cat ./coverage/lcov.info | coveralls"
```
Comment (physikerwelt) [inline package.json]: Delete

File: test-files/sample-mml.html (hunk 4 above)
Comment (physikerwelt) [inline test-files/sample-mml.html]: OK . I wont't change anything here for now

File: test/README.md (hunk 5 above)
Comment (physikerwelt) [inline test/README.md]: Delte.

File: test-files/sample-mml.html (hunk 4 above)
Comment (dpvc) [inline test-files/sample-mml.html]: The files in the `test-files` directory were just sample files that I used to test various things with the command-line tools.  This one is indeed a copy of the file from the MathJax website originally.  None of them are special, and any can be removed.  They are not referenced specifically anywhere but were just there to try out with the tools in `bin`.

File: test/features/math/simple.js (hunk 7 above)
Comment (zorkow) [inline test/features/math/simple.js]: What's the need for speaktext in response, if there is already alttext in the mml?

**Comment by pkra** [comment]
Thanks again for your help, @physikerwelt.

@mathjax/team-mathjax unless there are objections, I think it's a good starting point to merge.

**Comment by pkra** [comment]
Sorry this took us far longer than I thought. 

We are extremely grateful that you pushed us forward on this!

But we decided to go with a lightweight testing solution (tape). So I'm closing this in favor of https://github.com/mathjax/MathJax-node/pull/114.

Thank you again for all your help, @physikerwelt, and especially for making us figure out what we want to do here. We really appreciate it.
